### PR TITLE
T:26 Visual Studio Unit Test Project

### PR DESCRIPTION
**Pull request by ozdemirfadime**
added decryption_test_invalid method

### CODE REVIEW

File: projects/vs/psa-boot-test/psa-boot-test.cpp [hunk 1]
```diff
@@ -9,8 +9,8 @@ namespace psaboottest
 {
 	TEST_CLASS(psaboottest)
 	{
-		#define GET_TEST_PACKAGE(_pack)				(boot_upgrade_package_t*)copy_package(_pack, sizeof(_pack))
-		#define RELEASE_TEST_PACKAGE(_pack)			free(_pack)
+			#define GET_TEST_PACKAGE(_pack)				(boot_upgrade_package_t*)copy_package(_pack, sizeof(_pack))
```
Comment (zaya-mc) [inline projects/vs/psa-boot-test/psa-boot-test.cpp:12]: Please do not change unrelated lines, so revert these lines

File: projects/vs/psa-boot-test/psa-boot-testdata.h [hunk 2]
```diff
@@ -27,4 +27,45 @@ static const uint8_t td_valid_package[] = {
...
+    // Signature
+    0xBF, 0xA6, 0x82, 0x07, 0x89, 0x5E, 0x18, 0xB7, 0x77, 0xC9, 0xD3, 0x76, 0x62, 0x85, 0x61, 0xA6,
+    0xBE, 0x53, 0xEE, 0xAC, 0xCA, 0x12, 0x9B, 0x59, 0xBC, 0x11, 0x5F, 0xC2, 0xB4, 0x36, 0x95, 0x06,
+    0x44, 0xB1, 0xC5, 0xF9, 0xEF, 0xC6, 0x14, 0xD7, 0x88, 0xD9, 0x7D, 0x55, 0x0E, 0xFE, 0x18, 0x70,
+    0x2D, 0xE9, 0xD8, 0xCD, 0x7E, 0x96, 0xFF, 0xF0, 0xCE, 0x89, 0x2C, 0x38, 0xB3, 0xD0, 0x17, 0x88,
+    // Encrypted Image
+    0x0F, 0x66, 0xEF, 0x6F, 0x66, 0x34, 0xF5, 0x4C, 0x3B, 0x56, 0x7D, 0x4B, 0x24, 0x6A, 0xA4, 0x23,
+    0x73, 0xD8, 0xA0, 0xD9, 0x74, 0xD8, 0x59, 0x70, 0x02, 0x43, 0x32, 0x3F, 0xBC, 0xD2, 0xBB, 0x0B
+};
+static const uint8_t td_invalid_IV_package[] = {
```
Comment (zaya-mc) [inline projects/vs/psa-boot-test/psa-boot-testdata.h:50]: I dont see a test case for this package, are you planning to add a test case?

**Review by zaya-mc — CHANGES_REQUESTED** [review]

File: projects/vs/psa-boot-test/psa-boot-test.cpp [hunk 3]
```diff
@@ -5,72 +5,170 @@
...
+                /* Let us decrypt the valid package */
+                boot_decrypt_upgrade_package(package);
+                isEqual = arrays_are_equal(package->image, td_valid_package_plain_image_prefix_content, td_valid_package_plain_image_prefix_len);
+                /* After the encryption, the image must not be the same with expected plain image */
+                Assert::IsFalse(isEqual);
+
+            }
+            RELEASE_TEST_PACKAGE(package);
+        }
+        TEST_METHOD(decryption_test_invalid_size)
```
[the remark this answers is not in the document]
Comment (ozdemirfadime) [reply projects/vs/psa-boot-test/psa-boot-test.cpp:148]: This comment fixed.

File: projects/vs/psa-boot-test/psa-boot-test.cpp [hunk 4]
```diff
@@ -5,72 +5,170 @@
...
+{
+    return memcmp(arr1, arr2, len) == 0;
+}
+
+namespace PSABootTests
+{
+    TEST_CLASS(PositiveCases)
+    {
+    public:
+        /// <summary>
```
[the remark this answers is not in the document]
Comment (ozdemirfadime) [reply projects/vs/psa-boot-test/psa-boot-test.cpp:49]: done

File: projects/vs/psa-boot-test/psa-boot-test.cpp (hunk 1 above)
Comment (ozdemirfadime) [reply projects/vs/psa-boot-test/psa-boot-test.cpp]: done

File: projects/vs/psa-boot-test/psa-boot-test.cpp [hunk 5]
```diff
@@ -5,72 +5,170 @@
...
+                /* Let us decrypt the valid package */
+                boot_decrypt_upgrade_package(package);
+                isEqual = arrays_are_equal(package->image, td_valid_package_plain_image_content, td_valid_package_plain_image_len);
+                /* After the encryption, the image must not be the same with expected plain image */
+                Assert::IsFalse(isEqual);
+            }
+            RELEASE_TEST_PACKAGE(package);
+        }
+
+        TEST_METHOD(decryption_test_invalid_prefix)
```
[the remark this answers is not in the document]
Comment (ozdemirfadime) [reply projects/vs/psa-boot-test/psa-boot-test.cpp:132]: someone else needs to replace the prefix, does it need to again replace the prefix in the boot.c tests?

File: projects/vs/psa-boot-test/psa-boot-testdata.h (hunk 2 above)
Comment (ozdemirfadime) [reply projects/vs/psa-boot-test/psa-boot-testdata.h]: we use this package on the test case that is decryption_test_invalid_IV method

File: projects/vs/psa-boot-test/psa-boot-test.cpp [hunk 6]
```diff
@@ -5,72 +5,121 @@
...
+/// <summary>
+/// Copies a package into Read/Write Buffer. 
+/// Example Tests packages are constant values so
+/// decryption cannot update the packages, therefore,
+/// we need to copy the package to a updatable memory.
+/// </summary>
+/// <param name="source">Source Package</param>
+/// <param name="size">Source Package Size</param>
+/// <returns>Destination Package</returns>
+uint8_t* copy_package(const uint8_t* source, uint32_t size)
```
Comment (IBeRyUS) [inline projects/vs/psa-boot-test/psa-boot-test.cpp:20]: I think this function should return void*

File: projects/vs/psa-boot-test/psa-boot-test.cpp [hunk 7]
```diff
@@ -5,72 +5,121 @@
...
+{
+    uint8_t* dest = (uint8_t*)malloc(size);
+    if (dest == NULL)
```
Comment (IBeRyUS) [inline projects/vs/psa-boot-test/psa-boot-test.cpp:23]: I suggest optimising like this:
```cpp
if (NULL != dest)
{
    memcpy(dest, source, size);
}
return dest;
```

File: projects/vs/psa-boot-test/psa-boot-test.cpp [hunk 8]
```diff
@@ -5,72 +5,121 @@
...
+    return dest;
+}
+
+/// <summary>
+/// Compares two strings and returns true if strings are equals
+/// </summary>
+/// <param name="str1">String 1</param>
+/// <param name="str2">String 2</param>
+/// <returns>True if strings are equal, otherwise returns false</returns>
+bool arrays_are_equal(const uint8_t* arr1, const uint8_t* arr2, uint32_t len)
```
Comment (IBeRyUS) [inline projects/vs/psa-boot-test/psa-boot-test.cpp:39]: Naming suggestion is_arrays_are_equal. 
For me arrays_are_equal is not comparison it is directly equalising two arrays.

File: projects/vs/psa-boot-test/psa-boot-testdata.h [hunk 9]
```diff
@@ -4,7 +4,7 @@
 
 #include <stdint.h>
 
-static const uint8_t td_valid_package[] = {
+static const uint8_t td_package_valid[] = {
```
Comment (IBeRyUS) [inline projects/vs/psa-boot-test/psa-boot-testdata.h:7]: Why did you declare variables in header file?
Header files shouldn't include variable declarations.

**Review by IBeRyUS — CHANGES_REQUESTED** [review]

File: projects/vs/psa-boot-test/psa-boot-testdata.h (hunk 9 above)
Comment (zaya-mc) [reply projects/vs/psa-boot-test/psa-boot-testdata.h:7]: Actually, I have created this file. You are right, just this file is not an interface that hides data, instead, it keeps public test data which is tested by the test code. It is public information anyone can access. 
The idea was that separate the test data and the test code.

File: projects/vs/psa-boot-test/psa-boot-test.cpp (hunk 8 above)
Comment (zaya-mc) [reply projects/vs/psa-boot-test/psa-boot-test.cpp:39]: Make sense. Let fix it. 
`are_arrays_equal()`

File: projects/vs/psa-boot-test/psa-boot-testdata.h (hunk 9 above)
Comment (IBeRyUS) [reply projects/vs/psa-boot-test/psa-boot-testdata.h:7]: It's ok if only included by one file but still not good practice. I think making this proper C way is better. Move all of these declarations to psa-boot-testdata.c file and extern them in this header file.

Comment (zaya-mc) [reply projects/vs/psa-boot-test/psa-boot-testdata.h:7]: I don't see a problem, we could keep the test data inside the test source file to access all attributes(content/sizeof) of the test data, "extern" may mask some of them. 

Some C Test frameworks do the same thing; for example PSA Architecture Test  (please see the test_data.h)
https://github.com/ARM-software/psa-arch-tests/tree/master/api-tests/dev_apis/crypto/test_c002

It is not the bootloader implementation itself, it is just a private resource for the test code.

File: projects/vs/psa-boot-test/psa-boot-test.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (zaya-mc) [reply projects/vs/psa-boot-test/psa-boot-test.cpp:132]: @ozdemirfadime leave it with me, just get rid of this test case for now. I will add later and I will also change the PREFIX as "PSA".

File: projects/vs/psa-boot-test/psa-boot-testdata.h (hunk 2 above)
Comment (zaya-mc) [reply projects/vs/psa-boot-test/psa-boot-testdata.h]: It was an obsolete message, so ignore it.

File: projects/vs/psa-boot-test/psa-boot-test.cpp [hunk 10]
```diff
@@ -5,72 +5,222 @@
...
+        /// that is signed with the signing private key.
+        /// The image must be authenticated 
+        /// using the signing public key. 
+        /// </summary>
+        TEST_METHOD(authenticate_test_valid_image)
+        {
+            boot_upgrade_package_t* package = GET_TEST_PACKAGE(td_package_valid);
+            {
+                /* Let us the authenticate valid package */
+               bool isEqual = boot_authenticate_upgrade_package(package);
```
Comment (zaya-mc) [inline projects/vs/psa-boot-test/psa-boot-test.cpp:85]: please add a space before "bool" for right indentation.

File: projects/vs/psa-boot-test/psa-boot-test.cpp [hunk 11]
```diff
@@ -5,72 +5,222 @@
...
+                Assert::IsTrue(isEqual);
+            }
+            RELEASE_TEST_PACKAGE(package);
+
```
Comment (zaya-mc) [inline projects/vs/psa-boot-test/psa-boot-test.cpp:89]: Remove the empty space

File: projects/vs/psa-boot-test/psa-boot-test.cpp [hunk 12]
```diff
@@ -5,72 +5,222 @@
...
+        /// that has an invalid signature.
+        /// The image must not be authenticated 
+        /// using the signing public key. 
+        /// </summary
+        TEST_METHOD(authenticate_test_invalid_signature)
+        {
+            boot_upgrade_package_t* package = GET_TEST_PACKAGE(td_package_invalid_signature);
+            {                
+                /* Let us the valid package */
+                boot_authenticate_upgrade_package(package);
```
Comment (zaya-mc) [inline projects/vs/psa-boot-test/psa-boot-test.cpp:179]: You dont get and check the return value of boot_authenticate_upgrade_package(). 
You need to check only the return value here like authenticate_test_valid_image() test case above.

File: projects/vs/psa-boot-test/psa-boot-test.cpp [hunk 13]
```diff
@@ -5,72 +5,222 @@
...
+        TEST_METHOD(authentication_test_wrong_size)
+        {
+            boot_upgrade_package_t* package = GET_TEST_PACKAGE(td_package_wrong_size);
+            {
+                bool isEqual = arrays_are_equal(package->image, td_valid_package_plain_image_content, td_valid_package_plain_image_len);
+                /* We should have encrypted image so, strings should not be the same before decryption */
+                Assert::IsFalse(isEqual);
+                /* Let us decrypt the valid package */
+                boot_authenticate_upgrade_package(package);
+                isEqual = arrays_are_equal(package->image, td_valid_package_plain_image_content, td_valid_package_plain_image_len);
```
Comment (zaya-mc) [inline projects/vs/psa-boot-test/psa-boot-test.cpp:201]: You dont get and check the return value of boot_authenticate_upgrade_package().
You need to check only the return value here like authenticate_test_valid_image() test case above.

File: projects/vs/psa-boot-test/psa-boot-test.cpp [hunk 14]
```diff
@@ -5,72 +5,222 @@
...
+                /* After the encryption, the image must not be the same with expected plain image */
+                Assert::IsFalse(isEqual);
+            }
+            RELEASE_TEST_PACKAGE(package);
+        }
+        TEST_METHOD(authentication_test_invalid_image)
```
Comment (zaya-mc) [inline projects/vs/psa-boot-test/psa-boot-test.cpp:207]: Well done, that what we need for all invalid test cases.

**Review by zaya-mc — CHANGES_REQUESTED** [review]

File: source/boot.c [hunk 15]
```diff
@@ -75,6 +75,8 @@ int main(void) {
 
         boot_upgrade_package_t* upgrade_package = (boot_upgrade_package_t*)package_offset;
 
+        upgrade_package->metadata.prefix == "ZAYA";
```
Comment (zaya-mc) [inline source/boot.c:78]: I dont understand this line, Prefix already comes from Package.

**Review by zaya-mc — CHANGES_REQUESTED** [review]

File: projects/vs/psa-boot-test/psa-boot-test.cpp (hunk 10 above)
Comment (zaya-mc) [reply projects/vs/psa-boot-test/psa-boot-test.cpp:85]: @ozdemirfadime you haven't fixed that. Please fix.

File: projects/vs/psa-boot-test/psa-boot-test.cpp (hunk 11 above)
Comment (zaya-mc) [reply projects/vs/psa-boot-test/psa-boot-test.cpp:89]: Please fix. 
In general, please be consistent for all all source file formatting. If you put empty line in one case, please apply for all, Or; never add empty line.

File: projects/vs/psa-boot-test/psa-boot-test.cpp [hunk 16]
```diff
@@ -5,72 +5,215 @@
...
+                boot_decrypt_upgrade_package(package);
+
+                isEqual = arrays_are_equal(package->image, td_valid_package_plain_image_content, td_valid_package_plain_image_len);
+
+                /* After the encryption, the image must not be the same with expected plain image */
+                Assert::IsFalse(isEqual);
+            }
+            RELEASE_TEST_PACKAGE(package);
+        }
+        /// <summary>
```
Comment (zaya-mc) [inline projects/vs/psa-boot-test/psa-boot-test.cpp:121]: Please add an empty lines between all test cases. Thanks.

File: projects/vs/psa-boot-test/psa-boot-test.cpp [hunk 17]
```diff
@@ -5,72 +5,215 @@
...
+        {
+            boot_upgrade_package_t* package = GET_TEST_PACKAGE(td_package_wrong_size);
+            {
+                /* Let us the authenticate wrong size  package */
+                bool isEqual = boot_authenticate_upgrade_package(package);
+                Assert::IsFalse(isEqual);
+            }
+            RELEASE_TEST_PACKAGE(package);
+        }
+        TEST_METHOD(authentication_test_invalid_image)
```
Comment (zaya-mc) [inline projects/vs/psa-boot-test/psa-boot-test.cpp:200]: Please add a function header here like others.

File: source/boot.c (hunk 15 above)
Comment (zaya-mc) [reply source/boot.c:78]: Please remove this two lines.

**Review by zaya-mc — CHANGES_REQUESTED** [review]

**Review by zaya-mc — APPROVED** [review]
